File: TrackingVisualizer/Logger.py
from typing import Union
from enum import Enum
import time

class Logger:
    path: Union[str, None] = None
    
    class LogLevel(Enum):
        LOG_INFO = 1
        LOG_WARNING = 2
        LOG_ERROR = 3
        
    @staticmethod
    def set_file_path(path: str):
        Logger.path = path
        
    @staticmethod
    def emit_line(line: str):
        print(line)
        
        if Logger.path is not None:
            try:
                with open(Logger.path, "a") as file:
                    file.write(line + "\n")
            except FileNotFoundError:
                pass
    
    @staticmethod
    def log(level: LogLevel, message: str, *args):
        timestamp = time.strftime("(%Y-%m-%d %H:%M:%S)")
        
        if level == Logger.LogLevel.LOG_INFO:
            log_level = "INFO"
        elif level == Logger.LogLevel.LOG_WARNING:
            log_level = "WARNING"
        elif level == Logger.LogLevel.LOG_ERROR:
            log_level = "ERROR"
        else:
            log_level = "UNKNOWN"
            
        line = f"{timestamp} [{log_level}] {message.format(*args)}"
            
        Logger.emit_line(line)
        
            
    @staticmethod
    def log_info(message: str, *args):
        Logger.log(Logger.LogLevel.LOG_INFO, message, *args)
        
    @staticmethod
    def log_warning(message: str, *args):
        Logger.log(Logger.LogLevel.LOG_WARNING, message, *args)
        
    @staticmethod
    def log_error(message: str, *args):
        Logger.log(Logger.LogLevel.LOG_ERROR, message, *args)
        
    @staticmethod
    def append_info(message: str, *args):
        formated_message = message.format(*args)
        Logger.emit_line(" " * 24 + " - " + formated_message)

File: TimeTracker/src/Utils/Logger.cpp
#include "Logger.h"

LogLevel Logger::_log_level = LogLevel::LOG_INFO;

std::wstring Logger::_file_path = L"";

std::mutex Logger::_mutex{};

void Logger::set_file_path(const std::wstring& file_path)
{
	_file_path = file_path;

	std::wcout << "Logger file path set to: " << _file_path << std::endl;
}

void Logger::set_log_level(LogLevel level)
{
	_log_level = level;

	std::string log_level_str;
	switch (_log_level)
	{
	case LogLevel::LOG_INFO:
		log_level_str = "INFO";
		break;
	case LogLevel::LOG_WARNING:
		log_level_str = "WARNING";
		break;
	case LogLevel::LOG_ERROR:
		log_level_str = "ERROR";
		break;
	}

	DEBUG_LOG_LINE("Logger log level set to: " + log_level_str);
}

std::string Logger::_get_time_stamp()
{
	time_t raw_time;
	time(&raw_time);

	std::tm time_info;
	localtime_s(&time_info, &raw_time);

	char buffer[80];
	strftime(buffer, sizeof(buffer), "(%Y-%m-%d %H:%M:%S)", &time_info);

	return std::string(buffer);
}